controller/api/destination/watcher/endpoints_watcher.go

```diff
@@ -673,11 +977,63 @@ func diffAddresses(oldAddresses, newAddresses AddressSet) (add, remove AddressSe
 		}
 	}
 	add = AddressSet{
-		Addresses: addAddesses,
+		Addresses: addAddresses,
 		Labels:    newAddresses.Labels,
 	}
 	remove = AddressSet{
 		Addresses: removeAddresses,
 	}
 	return
 }
+
+func isTargetPortInSlice(targetPort namedPort, slice *discovery.EndpointSlice) bool {
+	if slice.Ports == nil && slice.Endpoints == nil {
+		return true
+	}
+
+	for _, p := range slice.Ports {
+		switch targetPort.Type {
+		case intstr.Int:
+			if *p.Port == targetPort.IntVal {
+				return true
+			}
+		case intstr.String:
+			if *p.Name == targetPort.StrVal {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func getEndpointSliceServiceID(es *discovery.EndpointSlice) (ServiceID, error) {
+	err := isValidSlice(es)
```

Comment:
up to you, but since this is the only place this is called, you could inline it here and that would remove some of the error handling boilerplate.